### ISSUE TITLE
Compilation error in KotlinGenerator generated code, when OTHER data type columns have a default value

### ISSUE THREAD

**Issue by dsvensson** [jOOQ/jOOQ#10900]
The following column (PostgreSQL):

```
 foo            | tsvector                 |           | not null | to_tsvector(''::text)
```

Produces:

```kotlin
@java.lang.Deprecated
val FOO: TableField<FooRecord, Any?> = createField(DSL.name("foo"), org.jooq.impl.DefaultDataType.getDefaultDataType("\"pg_catalog\".\"tsvector\"").nullable(false).defaultValue(DSL.field("to_tsvector(''::text)", org.jooq.impl.SQLDataType.OTHER)), this, "")
```

Which yields warning:

> Overload resolution ambiguity. All these functions match.
> `public abstract fun defaultValue(defaultValue: Any!): DataType<Any!> defined in org.jooq.DataType`
> `public abstract fun defaultValue(defaultValue: Field<Any!>!): DataType<Any!> defined in org.jooq.DataType`

The following generated Java code works:
```java
@java.lang.Deprecated
public final TableField<FooRecord, Object> FOO = createField(DSL.name("foo"), org.jooq.impl.DefaultDataType.getDefaultDataType("\"pg_catalog\".\"tsvector\"").nullable(false).defaultValue(DSL.field("to_tsvector(''::text)", org.jooq.impl.SQLDataType.OTHER)), this, "");
```

So due to `DataType<Object>` != `DataType<Any?>`  ...?

Even better would ofc be if it was properly supported, which borders to #5507 and #2938, but perhaps possible that other cases produces `SQLDataType.OTHER` and hit the same issue.

**Comment by lukaseder** [comment]
Thanks for your report. Yes, tsvector would be a nice type to support, but any other vendor specific  data type will produce the same issue, and it shouldn't cause a compilation error.

As a workaround, we could:

- Avoid the default type in this case
- Explicitly cast the default value to `Any` and "fix" the wrong overload resolution at runtime (this wouldnt' be the first time we do this, see e.g. https://stackoverflow.com/q/5361513/521799)
- Perhaps there's another way? Let's ask the experts: https://stackoverflow.com/q/64759424/521799

**Comment by lukaseder** [comment]
An auxiliary function can help. While this doesn't compile:

```kotlin
fun main() {
    class A<T> {
        fun m(t: T): Unit {
            print("T")
        }
        fun m(t: List<T>): Unit {
            print("List<T>")
        }
    }
    
    val a: A<Any> = A()
    val l: List<Any> = listOf()
    a.m(l)
}
```

This does:

```kotlin
fun <T> A<T>.mList(l: List<T>) {
    this.m(l);
}
a.mList(l)
```

That function would have to be generated, so I'm hoping there's a better way to resolve the ambiguity

**Comment by lukaseder** [comment]
No answer to the SO question yet. Without knowing what the correct approach should be here (and the workarounds being a bit dubious), I'm still postponing the fix.

The workaround is to implement a data type binding for the `tsvector` data type, or to exclude the column in the code generator configuration: https://www.jooq.org/doc/latest/manual/code-generation/codegen-advanced/codegen-config-database/codegen-database-includes-excludes/

**Comment by lukaseder** [comment]
The SO question (https://stackoverflow.com/q/64759424/521799) didn't conclude with a more satisfactory answer. There are two options:

- Avoid the overload by adding an unambiguous method like `A<T>.mList(l: List<T>)`
- Cast to `Any`.

I'm choosing the latter. It's the less nice solution, but has less impact on the code generator logic. We're already making sure behind the scenes that the two overloads behave the same way when users accidentally pick the wrong one, sending a `Field<T>` value to the `T` overload.

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.15.0 and 3.14.4 (#10946). Thanks again for your report!
